### ISSUE TITLE
`clippy::clone_on_copy` following recommendation on lock guards can lead to dead locks

### ISSUE THREAD

**Issue by juliusl** [rust-lang/rust-clippy#12957]
### Summary

Consider a RwLock on a Clone and/or Copy type.

```rs
static VAL: std::sync::RwLock<bool> = std::sync::RwLock::new(false);
```

Typically, when acquiring a read lock the caller should try to drop the read guard as soon as possible. So on types that are cloneable it's straightforward to call `.clone()`.

When fixing a bunch of lint issues and seeing the calling clone_on_copy lint, the easiest fix is to remove `.clone()`, which will compile. However, this can result in a deadlock since the previous guard will no longer be dropped without an explicit dereference.

Is it possible for the lint to skip types that are `Drop + Copy` or `Drop + Clone`?

### Additional Labels

@rustbot label + `l-suggestion-causes-error`

**Comment by Alexendoo** [comment]
The suggestion here seems correct https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=79443d91515a74b5e3e6083941febfc9

```
help: try dereferencing it: `*VAL.write().unwrap()`
```

**Comment by nk9** [comment]
If the suggestion is correct, should this be closed?

@rustbot label S-triage
